### PR TITLE
Use non-branching Montgomery ladder for point multiplication

### PR DESCRIPTION
**Pull request by aztech-dev**
Big thanks to @isislovecruft and @conradoplg for their advices (see #16)

@btcdrak @petertodd Gonna need a thorough review of this (build passes, except HHVM, but that's due to a missing function in their GMP implementation) -- the more reviewers, the merrier

### CODE REVIEW

**Comment by btcdrak** [comment]
ping @afk11

**Comment by btcdrak** [comment]
PR will need rebase.

**Comment by aztech-dev** [comment]
I will. For status, I've nearly got the implementation correct, but not all tests pass, I think I have a minor bug in there, but haven't had time to find it yet.

**Comment by btcdrak** [comment]
What is the status on this PR?
It also needs rebased

**Comment by aztech-dev** [comment]
Still the same... I'll rebase for starters, I have some freetime over the weekend, I'll try to wrap it up

**Comment by aztech-dev** [comment]
So I apparently got conditional swap working -- but it just takes too long (like really too long -- even with Gmp) with big numbers, I'll now work on optimizing it. Otherwise it's just not usable (or there is a bug somewhere)

**Comment by aztech-dev** [comment]
Ok, all tests should pass with this push, but BCMath is just... should we keep supporting it ? cc @afk11 @rubensayshi @btcdrak

**Comment by afk11** [comment]
Wow, I started the tests with BCMath ages ago, it's still running.. I am actually `time`ing it, so should be interesting :) 

The conditional swap seems to be in the UnsafePoint class, is it needed there? I figured Point would have the it, but it seems empty? 

I'm not sure, it's definitely a good idea to use the safer point class by default, but this is a pretty big issue for BcMath users..

**Comment by aztech-dev** [comment]
I created the unsafe point class to be able to swap the member values without throwing exceptions... I may be able to do without now that conditional swap works correctly, I'll do some tests.

And for BCMath, I think that the part that's taking long is base conversion used by the conditional swap (there's no native implementation available), though I have not benchmarked anything, this is just gut feeling.

**Comment by aztech-dev** [comment]
(and honestly, I haven't had the patience to run the full testsuite with BCMath :D )

**Comment by aztech-dev** [comment]
Allright, so I was able to get rid of the UnsafePoint class. For multiplication with BcMath, what I did is use the Montgomery ladder version that I had previously (the one where the array index can be leaked) -- I don't see any better way, BcMath is just too damn slow. If you see another solution, let me know...

**Comment by btcdrak** [comment]
Get rid of BCMath. There is no need for it when people can just install GMP.

**Comment by aztech-dev** [comment]
Allright, I'll do it when I have a moment.

**Comment by btcdrak** [comment]
knock knock :)

**Comment by aztech-dev** [comment]
Yup yup... was planning to do it today or tomorrow

**Comment by aztech-dev** [comment]
Well, I think it's finally ready... lest @conradoplg says my implementation is wrong again :D

**Comment by afk11** [comment]
Looks great, happy days! Tests pass here, though one CS nit.

File: src/Point.php [hunk 1]
```diff
@@ -280,38 +302,37 @@ public function getDouble()
...
-        return $this->adapter;
+        return [
+            'x' => (string) $this->x,
+            'y' => (string) $this->y,
+            'z' => (string) $this->order,
+            'curve' => $this->curve
+        ];
     }
-}
+}
```
Comment (afk11) [inline src/Point.php]: New line at EOF

Comment (aztech-dev) [inline src/Point.php]: Odd, I was sure I had ran a php-cs-fixer pass...

**Comment by btcdrak** [comment]
@conradoplg What do you think of this PR now?

**Comment by conradoplg** [comment]
Seems OK!

**Comment by aztech-dev** [comment]
Good news :) Glad to hear it

**Pull request merged**